[cmake][linux] Several fixes

hudokkow opened the pull request:
These are the low hanging ones...

Pending:
- generate kodi-config.cmake / xbmc-config.cmake from *.in files and install

Problems:
- folders with spaces in path aren't copied to install folder correctly (they are correct in build folder and properly quoted in <code>cmake_install.cmake</code>). See screenshot below.
- lib/kodi/system/players/VideoPlayer/libdvd***-x86_64-linux.so are not installed and I can't find them under cmake build/ folder. Something wrong with the build?
- a bunch of binaries are different of the autotools generated ones
- still not able to start kodi in the install folder:

> hudo@hudo:~/dev/tmp/c/bin$ ./kodi -p
> appPath: /home/hudo/dev/tmp/d/lib/kodi 
> Unable to find path to Kodi data files!

BTW, why are we installing empty folders like <code>share/kodi/addons/library.*</code>?

![virtualbox_2016-04-06_14-58-45](https://cloud.githubusercontent.com/assets/516183/14319530/7b1854ee-fc09-11e5-9461-83e8e65dab69.png)

ping @fetzerch and @notspiff

hudokkow committed:
[cmake] Install copying.txt and version.txt

hudokkow committed:
[cmake][linux] Install correct icons

hudokkow committed:
[cmake][linux] kodi.desktop is the name

hudokkow committed:
[cmake][linux] Add missing symlinks

hudokkow committed:
[cmake][linux] Install missing *.cmake files

fetzerch commented on project/cmake/installdata/common/common.txt:
```diff
@@ -3,11 +3,14 @@ sounds/*
+system/IRSSmap.xml
 system/peripherals.xml
 system/playercorefactory.xml
+system/X10-Lola-IRSSmap.xml
```
X10-Lola-IRSSmap.xml and IRSSmap.xml are used in `project/cmake/installdata/windows/irss.txt`. If we need them everywhere, could you drop them there?

hudokkow replied:
I don't know if they are used at all under linux. I just want that <code>diff -qr autotools/ cmake/ | sort</code> returns a nice *nada*. :wink:

notspiff replied:
I didn't put them in windows only just for the heck of it. They are only
used on windows, that is, they are the same file, the lola one is an
example for a specific remote and wont be used by the installation as such.
This is keymaps for 'lirc for windows'.
7. apr. 2016 15:43 skrev "Cristiano A. Silva" <notifications@github.com>:

> In project/cmake/installdata/common/common.txt
> <https://github.com/xbmc/xbmc/pull/9568#discussion_r58872648>:
>
> >  system/peripherals.xml
> >  system/playercorefactory.xml
> > +system/X10-Lola-IRSSmap.xml
>
> I don't know if they are used at all under linux. I just want that diff
> -qr autotools/ cmake/ | sort returns a nice *nada*. [image: :wink:]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/9568/files/5b9d2d52b4166c3e2e00be21ebad70f596d9b4c6..d8f6977c16c01c2a53242814a9c7494e8e93eb99#r58872648>
>

akva2 replied:
system/python shall always be empty on linux so why install it. replicating old bugs is not an acceptable answer.

fetzerch commented:
Thanks, yes that's certainly an area where we have to clean up things a bit.
I didn't look much into the installation so far, but your changes look good to me!
If @notspiff likes, a second pair of eyes would certainly not hurt.

Empty folders shouldn't be installed, looks like a bug.
For the libraries, I'll need to check that. Does Kodi run from the build dir?

hudokkow committed:
[cmake] Fix stray '\' in exec_prefix and datadir

hudokkow commented:
https://github.com/xbmc/xbmc/pull/9568/commits/24bedf9a1eaa9496e59824d1533482447d79967d cures problems with spaces/parentheses/whatever in paths during install to target folder.
Is it correct or did I got lucky and doing something completely wrong?

btw, it also fixes
> -- Prefix: /home/hudo/dev/tmp/c
-- Libdir: /home/hudo/dev/tmp/c/lib
-- Bindir: /home/hudo/dev/tmp/c/bin
-- Includedir: /home/hudo/dev/tmp/c/include
-- Datarootdir: /home/hudo/dev/tmp/c/share
-- Datadir: **${datarootdir}**

and now shows
> -- Prefix: /home/hudo/dev/tmp/c
-- Libdir: /home/hudo/dev/tmp/c/lib
-- Bindir: /home/hudo/dev/tmp/c/bin
-- Includedir: /home/hudo/dev/tmp/c/include
-- Datarootdir: /home/hudo/dev/tmp/c/share
-- Datadir: **/home/hudo/dev/tmp/c/share**

notspiff commented:
this should be correct. i have lost the history of the code due to a
gazillion rebases, but i'm sure @wsnipex, who wrote that code, had a reason
for the \ at some point. there's no reason now.

>

hudokkow commented:
Found out why it doesn't run inside the target install folder. bin/kodi is wrong:

> --- kodi_a	2016-04-07 13:41:08.660258646 +0100
+++ kodi_c	2016-04-07 14:57:17.251425396 +0100
@@ -21,10 +21,10 @@
 APP=Kodi
 bin_name=kodi
 SAVED_ARGS="$@"
-prefix="/home/hudo/dev/tmp/a"
-exec_prefix="/home/hudo/dev/tmp/a"
-datarootdir="${prefix}/share"
-LIBDIR="${exec_prefix}/lib"
+prefix="/home/hudo/dev/tmp/c"
+exec_prefix="/home/hudo/dev/tmp/c"
+datarootdir="/home/hudo/dev/tmp/c/share"
+LIBDIR="/home/hudo/dev/tmp/c/lib"
 CRASHLOG_DIR=${CRASHLOG_DIR:-$HOME}
 USERDATA_DIR="${HOME}/.${bin_name}"

Now to fix the folder.

@notspiff, yeah noticed that 20m ago. .gitignore files

akva2 commented:
and sorry if i come off as unnecessarily snarky, i'm under a lot of pressure at work and this is procrastination on a high level. not fair to take it out on innocents. i know but bare with me ;)

hudokkow commented:
No problem Arne. :wink: Don't forget to have some :beers:

hudokkow committed:
[cmake] Remove duplicate header include

hudokkow committed:
[cmake] These are the headers you're looking for

wsnipex commented:
For ref, datarootdir was supposed to be a shell var and has worked as is back then. Quite a lot has changed since then, so its possible something broke, but its not obvious to me what

hudokkow committed:
[cmake] Fix for Kodi refusing to run outside build dir

hudokkow commented:
@popcornmix, you might want to test this with rpi. In particular https://github.com/xbmc/xbmc/pull/9568/commits/dfa81493d43232b3a560582934e9700995ea0355 solves the run from install dir issues you mentioned in https://github.com/xbmc/xbmc/pull/9340#issuecomment-196322812 and that I was also experiencing in linux.
It only cost me 4 or 5 days to nail that down... and I must have looked at it 300+ times.

@fetzerch, other OSes pathsetup.cmake also use xbmc/. Is this by design? I assume kodi/ is the main dir and we just symlink xbmc to it.

fetzerch commented:
Yes, good catch. Could you add a fix for the other platforms as well?

Do you have more fixes in the queue, otherwise we could merge this soon.
I'd also prefer to drop 2b0a33c5c1a49509f67f15f0eba22f256f4fa83f however.

hudokkow commented:
Dropped 2b0a33c and squashed what seemed logic.
I have other fixes on the way but I probably can't test them before the weekend so might as well get this in.
This was only tested on Linux. Hopefully other platforms are OK.

hudokkow committed:
[cmake] Fix core_find_git_rev logic

hudokkow committed:
[cmake] Print some more debugging info

hudokkow commented:
Added two more commits.

Fun fact: autotools and cmake don't like each other. If you try to build with cmake after you built with autotools, its gonna fail.

I thought I was being clever by issuing <code>export CCACHE_DIR=~/.ccache-cmake</code> at the start of cmake build but I continued to observe problems with the build.
Took me a few hours to finally understand what was going on and <code>git clean</code> the tree.

Its probably because the flags are different, config.h isn't being generated, etc.

notspiff commented:
it's because autotools does everything in-tree. it generates the buildsystems for stuff which are generated out-of-tree with cmake. with those in place, you cannot do an out-of-tree afterwards.

hudokkow commented:
Yeah, not fun.
Anyways, lesson learned.

fetzerch commented:
Changes look good to me, so feel free to merge.

hudokkow commented:
Thanks! Merging now.
Just so we don't duplicate work, my todo list (so far) includes:
- properly parsing version.txt (adapt https://github.com/xbmc/xbmc/blob/master/project/cmake/scripts/common/prepare-env.cmake#L1-L17) but that needs next bullet...
- unify variable names between addons and core (APP_ROOT vs CORE_SOURCE_DIR for instance) so we can reuse code
- find out why APP_NAME_UC isn't being set and causes kodi-config.cmake to not be configured right
- properly set flags (right now we only set -std=c++11 except for WIN32) using something like https://github.com/xbmc/xbmc/blob/master/project/cmake/addons/CMakeLists.txt#L148-L155 and project/cmake/platform/${PLATFORM}/defines.txt
- generate config.h from cmake (this is probably going to be more work than all others together)
- add some comments to the files along the way
- minors

A couple of questions:
- why are cpluff configure files being generated in-tree? It's the only dep tainting the tree
- why are we building libdvd*** static? autotools builds them shared
- what platforms already have jenkins cmake build support?
